sdk/state/close.go

```diff
@@ -0,0 +1,148 @@
+		AmountToResponder:          c.responderClaimAmount(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return txClose, nil
+}
+
+// ProposeClose proposes parameters for a close transaction to be submitted earlier.
+// This should be used when participants are in agreement on the final txClose parameters, but would
+// like to submit earlier than the original observation time.
+func (c *Channel) ProposeCoordinatedClose(observationPeriodTime time.Duration, observationPeriodLedgerGap int64) (CoordinatedClose, error) {
+	txCoordinatedClose, err := txbuild.Close(txbuild.CloseParams{
+		ObservationPeriodTime:      observationPeriodTime,
+		ObservationPeriodLedgerGap: observationPeriodLedgerGap,
+		InitiatorSigner:            c.initiatorSigner(),
+		ResponderSigner:            c.responderSigner(),
+		InitiatorEscrow:            c.initiatorEscrowAccount().Address,
+		ResponderEscrow:            c.responderEscrowAccount().Address,
+		StartSequence:              c.startingSequence,
+		IterationNumber:            c.latestCloseAgreement.IterationNumber,
+		AmountToInitiator:          c.initiatorClaimAmount(),
+		AmountToResponder:          c.responderClaimAmount(),
+	})
+	if err != nil {
+		return CoordinatedClose{}, err
+	}
+	txCoordinatedClose, err = txCoordinatedClose.Sign(c.networkPassphrase, c.localSigner)
+	if err != nil {
+		return CoordinatedClose{}, nil
+	}
+	return CoordinatedClose{
+		observationPeriodTime:      observationPeriodTime,
+		observationPeriodLedgerGap: observationPeriodLedgerGap,
+		closeSignatures:            txCoordinatedClose.Signatures(),
+	}, nil
+}
+
+func (c *Channel) ConfirmCoordinatedClose(cc CoordinatedClose) (CoordinatedClose, error) {
+	txCoordinatedClose, err := txbuild.Close(txbuild.CloseParams{
+		ObservationPeriodTime:      cc.observationPeriodTime,
+		ObservationPeriodLedgerGap: cc.observationPeriodLedgerGap,
+		InitiatorSigner:            c.initiatorSigner(),
+		ResponderSigner:            c.responderSigner(),
+		InitiatorEscrow:            c.initiatorEscrowAccount().Address,
+		ResponderEscrow:            c.responderEscrowAccount().Address,
+		StartSequence:              c.startingSequence,
+		IterationNumber:            c.latestCloseAgreement.IterationNumber,
+		AmountToInitiator:          c.initiatorClaimAmount(),
+		AmountToResponder:          c.responderClaimAmount(),
+	})
+	if err != nil {
+		return CoordinatedClose{}, err
+	}
+
+	// If remote has not signed coordinated close, error as is invalid.
+	signed, err := c.verifySigned(txCoordinatedClose, cc.closeSignatures, c.remoteSigner)
+	if err != nil {
+		return CoordinatedClose{}, err
+	}
+	if !signed {
+		return CoordinatedClose{}, fmt.Errorf("verifying coordinated close signed by remote: not signed by remote")
+	}
+
+	// If local has not signed, sign.
+	signed, err = c.verifySigned(txCoordinatedClose, cc.closeSignatures, c.localSigner)
+	if err != nil {
+		return CoordinatedClose{}, err
+	}
+	if !signed {
```

Comment:
So the general steps are:

```
// 1. Initiator or Responder submits latest declaration tx
// (in steps 2-4 Initiator and Responder are interchangeable, as long as they alternate)
// 2. Initiator calls ProposeCoordinatedClose
// 3. Responder calls ConfirmCoordinatedClose
// 4. Initiator calls ConfirmCoordinatedClose
```
So in step 4 initiator (local) has already signed. They sign it when they propose it.

This shouldn't be a bug though. If you're the initiator, when you propose the new `txClose` you sign it and send to the responder. If the responder submits it, great that's what you wanted. If responder doesn't, then you can still just wait for `O` to pass and submit the original `txClose` that you already have fully signed.

the coordinated close has the same data as the latest `txClose` just with the shortened `O` period

happy to discuss further

Reply:
that makes sense since it's ok for responder to submit earlier since this is a coordinated close.

wondering if it's already signed in step 2, then why do we need to check if it is !signed in step 4?
(basically I'm wondering if we are checking for a state that isn't possible. and if that state occurs then it's a programmer error). can discuss on call.